### PR TITLE
EVG-5166 refactor blocked state, handle display tasks

### PR DESCRIPTION
**Pull request by john-m-liu**

### CODE REVIEW

File: model/task/task.go [hunk 1]
```diff
@@ -37,6 +37,9 @@ const (
 
 	// length of time to cache the expected duration in the task document
 	predictionTTL = 15 * time.Minute
+
+	TaskBlocked = "blocked"
```
Comment (richardsamuels) [inline model/task/task.go]: nit, but these can be private

File: public/static/js/mci_module.js [hunk 2]
```diff
@@ -292,7 +292,7 @@ var mciModule = angular.module('MCI', [
   return function(task) {
     if (task.status == 'started') {
       return 'started';
-    } else if (task.status == 'undispatched' && task.activated) {
+    } else if (task.status == 'undispatched' && task.activated || task.display_only && task.task_waiting) {
```
Comment (richardsamuels) [inline public/static/js/mci_module.js]: Are you trying to say  `(task.status == 'undispatched' && task.activated) || task.display_only && task.task_waiting)`
or `task.status == 'undispatched' && (task.activated || task.display_only && task.task_waiting)`? Please clarify with parens.

File: public/static/js/mci_module.js [hunk 3]
```diff
@@ -258,7 +258,7 @@ var mciModule = angular.module('MCI', [
       return task;
     }
     var cls = task.status;
-    if (task.status == 'undispatched') {
+    if (task.status == 'undispatched' || task.display_only && task.task_waiting) {
```
Comment (richardsamuels) [inline public/static/js/mci_module.js]: Same note about parens here.

File: model/task/task_test.go [hunk 4]
```diff
@@ -1085,3 +1085,48 @@ func TestGetTestResultsForDisplayTask(t *testing.T) {
...
+	t2 := Task{
+		Id:     "t2",
+		Status: evergreen.TaskFailed,
+		DependsOn: []Dependency{
+			{TaskId: "t3", Status: evergreen.TaskFailed},
+		},
+	}
+	assert.NoError(t2.Insert())
+	t3 := Task{
+		Id: "t3",
```
Comment (richardsamuels) [inline model/task/task_test.go]: t3 needs a status set.

File: model/task/task.go [hunk 5]
```diff
@@ -1632,3 +1635,62 @@ func (t *Task) GetJQL(searchProjects []string) string {
...
+		state, err := depTask.BlockedState()
+		if err != nil {
+			return "", err
+		}
+		if state == TaskBlocked {
+			return TaskBlocked, nil
+		} else if depTask.Status == evergreen.TaskSucceeded || depTask.Status == evergreen.TaskFailed {
+			if depTask.Status != dependency.Status && dependency.Status != AllStatuses {
+				return TaskBlocked, nil
+			}
```
Comment (richardsamuels) [inline model/task/task.go]: In the case where depTask.Status == evergreen.TaskSucceeded, and !(depTask.Status != dependency.Status && dependency.Status != AllStatuses), this function will return "", nil. Is this what you want?

File: model/task/task.go [hunk 6]
```diff
@@ -1632,3 +1635,62 @@ func (t *Task) GetJQL(searchProjects []string) string {
...
+		dependencyIDs = append(dependencyIDs, d.TaskId)
+	}
+	dependentTasks, err := Find(ByIds(dependencyIDs).WithFields(DisplayNameKey, StatusKey,
+		ActivatedKey, BuildVariantKey, DetailsKey, DependsOnKey))
+	if err != nil {
+		return "", errors.Wrap(err, "error finding dependencies")
+	}
+	taskMap := map[string]*Task{}
+	for _, t := range dependentTasks {
+		taskMap[t.Id] = &t
```
Comment (richardsamuels) [inline model/task/task.go]: You are taking the address of the task that was copied by the range operator here. This won't work as you expect.

File: service/task.go [hunk 7]
```diff
@@ -390,134 +384,35 @@ func getTaskDependencies(t *task.Task) ([]uiDep, string, error) {
...
-			}
-		}
-	}
-
-	idToDep := make(map[string]task.Task)
-	for _, dep := range dependencies {
-		idToDep[dep.Id] = dep
+	taskMap := map[string]*task.Task{}
+	for _, t := range dependencies {
+		taskMap[t.Id] = &t
```
Comment (richardsamuels) [inline service/task.go]: See my other comment about the range operator and addresses elsewhere.

File: model/task/task.go (hunk 1 above)
Comment (john-m-liu) [reply model/task/task.go]: done

File: model/task/task.go (hunk 6 above)
Comment (john-m-liu) [reply model/task/task.go]: fixed

File: model/task/task.go (hunk 5 above)
Comment (john-m-liu) [reply model/task/task.go]: Yes, that means the dependency has been satisfied

File: model/task/task_test.go (hunk 4 above)
Comment (john-m-liu) [reply model/task/task_test.go]: added

File: public/static/js/mci_module.js (hunk 3 above)
Comment (john-m-liu) [reply public/static/js/mci_module.js]: done

File: public/static/js/mci_module.js (hunk 2 above)
Comment (john-m-liu) [reply public/static/js/mci_module.js]: done

File: service/task.go (hunk 7 above)
Comment (john-m-liu) [reply service/task.go]: fixed

**Pull request merged**